src/backend/metal/src/device.rs

```diff
@@ -795,18 +809,67 @@ impl Device {
             });
         }
 
-        if self.shared.private_caps.argument_buffers {
+        if caps.argument_buffers {
             descriptor.set_support_argument_buffers(true);
         }
 
-        Ok(descriptor)
+        Some(descriptor)
     }
 
-    pub fn create_sampler_from_descriptor(
-        &self,
-        descriptor: &metal::SamplerDescriptorRef,
-    ) -> n::Sampler {
-        n::Sampler(self.shared.device.lock().new_sampler(descriptor))
+    fn make_sampler_data(info: &image::SamplerInfo) -> msl::SamplerData {
+        fn map_address(wrap: image::WrapMode) -> msl::SamplerAddress {
+            match wrap {
+                image::WrapMode::Tile => msl::SamplerAddress::Repeat,
+                image::WrapMode::Mirror => msl::SamplerAddress::MirroredRepeat,
+                image::WrapMode::Clamp => msl::SamplerAddress::ClampToEdge,
+                image::WrapMode::Border => msl::SamplerAddress::ClampToBorder,
+            }
+        }
+
+        let lods: Range<f32> = info.lod_range.start.into() .. info.lod_range.end.into();
+        msl::SamplerData {
+            coord: if info.normalized {
+                msl::SamplerCoord::Normalized
+            } else {
+                msl::SamplerCoord::Pixel
+            },
+            min_filter: match info.min_filter {
+                image::Filter::Nearest => msl::SamplerFilter::Nearest,
+                image::Filter::Linear => msl::SamplerFilter::Linear,
+            },
+            mag_filter: match info.mag_filter {
+                image::Filter::Nearest => msl::SamplerFilter::Nearest,
+                image::Filter::Linear => msl::SamplerFilter::Linear,
+            },
+            mip_filter: match info.min_filter {
+                image::Filter::Nearest if info.lod_range.end < image::Lod::from(0.5) =>
```

Comment:
Why do we need the special LOD logic here?

Reply:
same logic as we do in `make_sampler_descriptor`